src/Discord.Net.SlashCommands/SlashCommandService.cs

```diff
@@ -0,0 +1,675 @@
+
+            foreach (var command in existing)
+                props.Add(command.ToApplicationCommandProps());
+
+            return await ClientHelper.BulkOverwriteGuildApplicationCommand(Client, guild.Id, props.ToArray()).ConfigureAwait(false);
+        }
+
+        private void LoadModuleInternal (ModuleInfo module)
+        {
+            _moduleDefs.Add(module);
+
+            foreach (var command in module.SlashCommands)
+                _slashCommandMap.AddCommand(command, command.IgnoreGroupNames);
+
+            foreach (var command in module.ContextCommands)
+                _contextCommandMap.AddCommand(command, command.IgnoreGroupNames);
+
+            foreach (var interaction in module.Interactions)
+                _interactionCommandMap.AddCommand(interaction, interaction.IgnoreGroupNames);
+
+            foreach (var subModule in module.SubModules)
+                LoadModuleInternal(subModule);
+        }
+
+        /// <summary>
+        /// Remove a loaded module from <see cref="SlashCommandService.Modules"/>
+        /// </summary>
+        /// <param name="type"><see cref="SlashModuleBase{T}"/> that will be removed</param>
+        /// <returns></returns>
+        public async Task<bool> RemoveModuleAsync (Type type)
+        {
+            await _lock.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                if (!_typedModuleDefs.TryRemove(type, out var module))
+                    return false;
+
+                return RemoveModuleInternal(module);
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+        private bool RemoveModuleInternal (ModuleInfo moduleInfo)
+        {
+            if (!_moduleDefs.Remove(moduleInfo))
+                return false;
+
+
+            foreach (var command in moduleInfo.SlashCommands)
+            {
+                _slashCommandMap.RemoveCommand(command);
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Execute a slash command from a given <see cref="ISlashCommandContext"/>
+        /// </summary>
+        /// <param name="context">A command context that will be used to execute the command, <see cref="ISlashCommandContext.Interaction"/>
+        /// must be type of <see cref="SocketSlashCommand"/></param>
+        /// <param name="input">Command string that will be used to parse the <see cref="SlashCommandInfo"/>. Use the
+        /// <see cref="WebSocketExtensions.GetCommandKeywords(SocketSlashCommand)"/> to get the input </param>
+        /// <param name="services">Services that will be injected into the declaring type</param>
+        /// <returns>A task representing the command execution process, with an <see cref="IResult"/> containg the execution information as it result.</returns>
+        public async Task<IResult> ExecuteSlashCommandAsync (ISlashCommandContext context, string[] input, IServiceProvider services)
```

Comment:
The premis of the different execute functions is redundant, the checks can be done by the service to figure out what type of interaction it is and what command to call. Relying on the user to specify this makes the whole experience painful for them

Reply:
With the `string[] input` parameter seems really redundant and offloading the work to the user. The service is more than capable deserializing the context passed in to find the array if need be